### DISCUSSION TITLE
Making variants optional or required based on presence in `defaultVariants`

### DISCUSSION THREAD

**Discussion by nayaabkhan** [vanilla-extract-css/vanilla-extract#588] [Ideas]
Hi! Thanks for a great tool-set.

I've been deep into exploring the APIs and stumbled upon something that kinda feels abnormal.

Say:
```ts
export const heading = recipe({
  variants: {
    weight: {
      bold: atoms({ fontWeight: 'bold' }),
      semibold: atoms({ fontWeight: 'semibold' }),
    },
    size: {
      '1': atoms({ fontSize: 36 }),
      '2': atoms({ fontSize: 28 }),
      '3': atoms({ fontSize: 24 }),
    },
  },

  defaultVariants: {
    weight: 'bold',
  },
})

export type HeadingVariants = RecipeVariants<typeof heading>
```

I expected `HeadingVariants` to have `weight` as optional because a default is provided, but `size` to be required:

```ts
type HeadingVariants = {
  weight?: 'bold' | 'semibold'
  size: '1' | '2' | '3'
};
```

But we have all fields optional:

```ts
type HeadingVariants = {
  weight?: 'bold' | 'semibold'
  size?: '1' | '2' | '3'
};
```

- Is it something by design or not trivial to do with TypeScript?
- Is there a workaround/recommendation on how to achieve the expected outcome?

Thank a lot, in advance.

**Comment by graup** [comment]
It seems to me that this is by design. Variants are always optional – if not set, simply no extra style is added. If you want to make it required, just change the variants type, e.g. something like this (untested):

```ts
type OptionalHeadingVariants = RecipeVariants<typeof heading>;
export type HeadingVariants = OptionalHeadingVariants & Required<Pick<OptionalHeadingVariants, 'size'>>;
```

**Reply by HeartSquared** [reply]
Just tested this and the suggestion actually doesn't work:

<img width="1151" alt="image" src="https://user-images.githubusercontent.com/25891850/162345660-09b86967-7f39-4fbf-8707-5831028e1469.png">

**Reply by graup** [reply]
@HeartSquared coule you show the rest of your code? Maybe as a Typescript Playground link?
